src/board.ts

```diff
@@ -334,6 +335,111 @@ export function getKeyAtDomPos(pos: cg.NumberPair, asWhite: boolean, bounds: Cli
+  if (targetKey === undefined) return false;
+  const pos = key2pos(targetKey);
+  // +
+  if (orig[0] === pos[0] || orig[1] === pos[1]) return true;
+  // x
+  if (Math.abs(orig[0] - pos[0]) === Math.abs(orig[1] - pos[1])) return true;
+  // knight
+  if (Math.abs(orig[0] - pos[0]) === 2 && Math.abs(orig[1] - pos[1]) === 1) return true;
+  if (Math.abs(orig[0] - pos[0]) === 1 && Math.abs(orig[1] - pos[1]) === 2) return true;
+  // all other
+  return false;
+}
+
+function clamp(n: number) {
+  if (n < 0) return 0
+  if (n > 7) return 7
+  return n
+}
+
+interface SnapCriteria {
+  x1: number;
+  y1: number;
+  x2: number;
+  y2: number;
+  dd: number;
+}
+
+const snapFunctions = {
+  e: ({y1, x2}: SnapCriteria): cg.Pos => [clamp(x2), y1],
+  ne: ({x1, y1, dd}: SnapCriteria): cg.Pos => {const c = Math.min(Math.min(dd, 7 - x1), Math.min(dd, 7 - y1));return [x1 + c, y1 + c]},
+  n: ({x1, y2}: SnapCriteria): cg.Pos => [x1, clamp(y2)],
+  nw: ({x1, y1, dd}: SnapCriteria): cg.Pos => {const c = Math.min(Math.min(dd, x1), Math.min(dd, 7 - y1));return [x1 - c, y1 + c];},
+  w: ({y1, x2}: SnapCriteria): cg.Pos => [clamp(x2), y1],
+  sw: ({x1, y1, dd}: SnapCriteria): cg.Pos => {const c = Math.min(Math.min(dd, x1), Math.min(dd, y1));return [x1 - c, y1 - c];},
+  s: ({x1, y2}: SnapCriteria): cg.Pos => [x1, clamp(y2)],
+  se: ({x1, y1, dd}: SnapCriteria): cg.Pos => {const c = Math.min(Math.min(dd, 7 - x1), Math.min(dd, y1));return [x1 + c, y1 - c];},
+}
+
+const snapAngles = [
+  {angle: 0, snap:   (asWhite: boolean) => asWhite ? snapFunctions.e : snapFunctions.w},
+  {angle: 45, snap:  (asWhite: boolean) => asWhite ? snapFunctions.ne : snapFunctions.sw},
+  {angle: 90, snap:  (asWhite: boolean) => asWhite ? snapFunctions.n : snapFunctions.s},
+  {angle: 135, snap: (asWhite: boolean) => asWhite ? snapFunctions.nw : snapFunctions.se},
+  {angle: 180, snap: (asWhite: boolean) => asWhite ? snapFunctions.w : snapFunctions.e},
+  {angle: 225, snap: (asWhite: boolean) => asWhite ? snapFunctions.sw : snapFunctions.ne},
+  {angle: 270, snap: (asWhite: boolean) => asWhite ? snapFunctions.s : snapFunctions.n},
+  {angle: 315, snap: (asWhite: boolean) => asWhite ? snapFunctions.se : snapFunctions.nw},
+  {angle: 360, snap: (asWhite: boolean) => asWhite ? snapFunctions.e : snapFunctions.w},
+  {angle: 405, snap: (asWhite: boolean) => asWhite ? snapFunctions.e : snapFunctions.w},
+];
+const angleThresholds = [snapAngles[0]].concat(snapAngles.slice(0, -1)).map(({angle}, i, arr) => ({
+  angle: (angle + snapAngles[i].angle) / 2,
+  snap: arr[i].snap
+})).slice(1);
+
+export function getSnappedKeyAtDomPos(orig: cg.Key, pos: cg.NumberPair, asWhite: boolean, bounds: ClientRect): cg.Key | undefined {
+  // 1. Get key at dom pos
+  // 2. If move is valid, short circuit
+  // 3. If move is not valid, pick nearest angle and snap distance to dragged distance
+  // Only snapping to + and x currently.
+  // TODO: also snap to knight angles, but stop after going past knight tile?
```

Comment:
Should be simpler now: create another set of snapAngles / angleThresholds to use when origCenterToMouseCoordDistance is within knight tile range

Reply:
I tried it out, and it worked, but it I couldn't figure out a distance and angle combination I liked more than without snapping to knight angles.